http/peer-id-auth.md

```diff
@@ -0,0 +1,331 @@
+as the PeerID alone may not be enough to validate a signature. The Public Key is
+encoded per the [Peer ID spec] under the section "Keys" section.
+
+## Mutual Client and Server Peer ID Authentication
+
+The following protocol allows both the client and server to authenticate each
+other's Peer ID by having them each sign a challenge issued by the other. The
+protocol operates as follows:
+
+### Server Initiated Handshake
+
+1. The client makes an HTTP request to an authenticated resource.
+
+2. The server responds with status code 401 (Unauthorized) and set the header:
+   ```
+   WWW-Authenticate: libp2p-PeerID challenge-client="<challenge-string>", public-key="<base64-encoded-public-key-bytes>", opaque="<opaque-value>"
+   ```
+
+   The public-key parameter is the server's public key. It is the same public
+   key used to derive the server's peer id.
+
+   The opaque parameter is opaque to client. The client MUST return the opaque
+   parameter back to the server. The server MAY use the opaque parameter to
+   encode state.
+
+3. The client makes another HTTP request to the same authenticated resource and
+   sets the header:
+
+   ```
+   Authorization: libp2p-PeerID public-key="<base64-encoded-public-key-bytes>", opaque="<opaque-from-server>", challenge-server="<challenge-string>", sig="<base64-signature-bytes>"
+   ```
+
+   The public-key parameter is the client's public key. It is the same public
+   key used to derive the client's peer id.
+
+   The `sig` param represents a signature over the parameters:
+    - `challenge-client`
+    - `server-public-key` the bytes of the server's public-key encoded per the [Peer ID spec].
+    - `hostname`
+
+4. The server SHOULD verify the signature using the server name used in the TLS
+   session. The server MUST return 401 Unauthorized if the server fails to
+   validate the signature. If the signature is valid, the server has
+   authenticated the client's public key, and thus its PeerID. The server SHOULD
+   proceed to serve the HTTP request. The server MUST set the following response
+   headers:
+   ```
+   Authentication-Info: libp2p-PeerID, sig="<base64-signature-bytes>" bearer="<base64-encoded-opaque-blob>"
+   ```
+
+   The `sig` param represents a signature over the parameters:
+    - `challenge-server`
+    - `client-public-key` the bytes of the client's public-key encoded per the [Peer ID spec].
+    - `hostname`
+
+   The `bearer` token allows the client to make future Peer ID authenticated
+   requests. The value is opaque to the client, and the server may use it to
+   store authentication state such as:
+     - The client's Peer ID.
+     - The `hostname` parameter.
+     - The token creation date (to allow tokens to expire).
+
+5. The client MUST verify the signature. After verification the client has
+   authenticated the server's Peer ID. The client SHOULD send the `bearer`
+   token for Peer ID authenticated requests.
```

Comment:
If it's mutual auth, both the client and server need to exchange bearer tokens, right?

Reply:
the bearer token is to allow the server to statelessly authenticate the client. Clients are expected to store the state that they have authenticated the server.

Reply:
How would this work for browsers? Should they re-authenticate servers after a TTL?

Reply:
What are the issues specific to browsers?

On the Go client implementation, the client stores a map from hostname to
```
type tokenInfo struct {
	token      string
	insertedAt time.Time
	peerID     peer.ID
}
```

Reply:
I think this is fine. 

I thought it might cause an issue with the browser since they don't control the transport connection. But we don't do anything special in go either. 

Browsers can have an implementation defined `TokenTTL` similar to the go implementation.

Reply:
This is fine. Even if the connection is broken and a new connection is established after the auth is run. It'll be as secure as the WebPKI.